source/common/protobuf/utility.cc

```diff
@@ -403,80 +404,77 @@ void checkForDeprecatedNonRepeatedEnumValue(const Protobuf::Message& message,
 
     // If this field is not in use, continue.
-    if ((field->is_repeated() && reflection->FieldSize(message, field) == 0) ||
-        (!field->is_repeated() && !reflection->HasField(message, field))) {
-      continue;
+    if ((field.is_repeated() && reflection->FieldSize(message, &field) == 0) ||
+        (!field.is_repeated() && !reflection->HasField(message, &field))) {
+      return nullptr;
     }
 
     // If this field is deprecated, warn or throw an error.
-    if (field->options().deprecated()) {
+    if (field.options().deprecated()) {
       const std::string warning = absl::StrCat(
-          "Using {}deprecated option '", field->full_name(), "' from file ", filename,
+          "Using {}deprecated option '", field.full_name(), "' from file ", filename,
           ". This configuration will be removed from "
           "Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated "
           "for details.");
 
-      deprecatedFieldHelper(
-          runtime, true /*deprecated*/,
-          field->options().GetExtension(envoy::annotations::disallowed_by_default),
-          absl::StrCat("envoy.deprecated_features:", field->full_name()), warning, message);
+      deprecatedFieldHelper(runtime_, true /*deprecated*/,
+                            field.options().GetExtension(envoy::annotations::disallowed_by_default),
+                            absl::StrCat("envoy.deprecated_features:", field.full_name()), warning,
+                            message);
     }
+    return nullptr;
+  }
 
-    // If this is a message, recurse to check for deprecated fields in the sub-message.
-    if (field->cpp_type() == Protobuf::FieldDescriptor::CPPTYPE_MESSAGE) {
-      if (field->is_repeated()) {
-        const int size = reflection->FieldSize(message, field);
-        for (int j = 0; j < size; ++j) {
-          checkForUnexpectedFields(reflection->GetRepeatedMessage(message, field, j),
-                                   validation_visitor, runtime);
+  void onMessage(const Protobuf::Message& message, const void*) override {
+    // Reject unknown fields.
+    const auto& unknown_fields = message.GetReflection()->GetUnknownFields(message);
+    if (!unknown_fields.empty()) {
+      std::string error_msg;
+      for (int n = 0; n < unknown_fields.field_count(); ++n) {
+        if (unknown_fields.field(n).number() == ProtobufWellKnown::OriginalTypeFieldNumber) {
+          continue;
         }
-      } else {
-        checkForUnexpectedFields(reflection->GetMessage(message, field), validation_visitor,
-                                 runtime);
+        error_msg += absl::StrCat(n > 0 ? ", " : "", unknown_fields.field(n).number());
+      }
+      // We use the validation visitor but have hard coded behavior below for deprecated fields.
+      // TODO(htuch): Unify the deprecated and unknown visitor handling behind the validation
+      // visitor pattern. https://github.com/envoyproxy/envoy/issues/8092.
```

Comment:
Checking: you still have further cleanup plans?

Reply:
Yeah. This PR just refactors some proto traversal stuff. https://github.com/envoyproxy/envoy/issues/8092 tracks the more general "move the deprecated field error reporting and exception generation to the validation visitor callbacks", i.e. having an `onDeprecatedField()` callback alongside `onUnknownField()` in `ValidationVisitor`. This is mostly a mechanical code movement that shouldn't really change any existing behavior. @nezdolik FYI.